subsystems/ElevatorBase.java

```diff
@@ -1,5 +1,179 @@
 package org.frc5587.lib.subsystems;
 
-public abstract class ElevatorBase {
+import edu.wpi.first.math.controller.ElevatorFeedforward;
+import edu.wpi.first.math.controller.ProfiledPIDController;
+import edu.wpi.first.math.trajectory.TrapezoidProfile;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.motorcontrol.MotorController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.ProfiledPIDSubsystem;
+
+public abstract class ElevatorBase extends ProfiledPIDSubsystem {
+    protected ElevatorConstants constants;
+    protected MotorController motor;
+    protected ElevatorFeedforward ffController;
+    protected ProfiledPIDController pidController;
+    protected DigitalInput limitSwitch;
+
+    public static final class ElevatorConstants {
+        public final double gearing, rotationsToMeters;
+        public final double[] softLimits;
+        public final int zeroOffset, encoderCPR, switchPort;
+        public final boolean switchInverted;
+        public final ProfiledPIDController pid;
+        public final ElevatorFeedforward ff;
+
+        public ElevatorConstants(
+                double gearing, 
+                double rotationsToMeters,
+                double[] softLimits,
+                int zeroOffset,
+                int encoderCPR,
+                int switchPort,
+                boolean switchInverted,
+                ProfiledPIDController pid,
+                ElevatorFeedforward ff) {
+            this.gearing = gearing;
+            this.rotationsToMeters = rotationsToMeters;
+            this.softLimits = softLimits;
+            this.zeroOffset = zeroOffset;
+            this.encoderCPR = encoderCPR;
+            this.switchPort = switchPort;
+            this.switchInverted = switchInverted;
+            this.pid = pid;
+            this.ff = ff;
+        }
+    }
     
+    public ElevatorBase(ElevatorConstants constants, MotorController motor) {
+        super(constants.pid);
+        this.constants = constants;
+        this.motor = motor;
+        this.ffController = constants.ff;
+        this.limitSwitch = new DigitalInput(constants.switchPort);
```

Comment:
instead we should do 2 different constructors, one with a limitswitch, one without. Just make it a param on the constructor as an int

Reply:
just cause we want this to be dynamic and might do something slightly different than just one limit switch.

Reply:
meaning we should allow an option for multiple limit switches + the option for 0; or just the option for 0/1 limit switches?

Reply:
um actually an arraylist wouldnt work cause you wouldnt know which limit switch was which...

Reply:
wait wait, yes it would. make a constant thats a parallel array to the arraylist and each limit switch has its switchport, so you can tell based of the index of the array.

this makes no sense typing it out, we can talk about it later

Reply:
charlie i know how to do the code I just didn't know what you wanted

Reply:
ok :sob: thank u